Update image parsing implementation

bernardjkim opened the pull request:
## Description
<!--Required. Provide high-level overview of what the change is for.-->
Include image digest in docker images and update the parsing implementation to allow parsing of images with both a tag and a digest.

These changes will allow proper parsing of images in the format:
```
k8s.gcr.io/ingress-nginx/controller:v0.49.3@sha256:35fe394c82164efa8f47f3ed0be981b3f23da
```

## Type of change
<!--Required. Keep only those that apply.-->

* Bug fix (non-breaking change which fixes an issue)

## Linked tickets and other PRs
<!--Required. Keep only those that apply.-->

## TODOs
<!--Required. Keep only those that apply and check them off as they get completed.-->

- [x] Self-review the change
- [x] Write tests
- [x] Perform manual testing
- [x] Address review feedback

## Testing done
<!--Required. Explain what kind of testing these changes underwent.-->
- [x] Successfully parses `k8s.gcr.io/ingress-nginx/controller:v0.49.3@sha256:35fe394c82164efa8f47f3ed0be981b3f23da`

Install gravity cmd with `make install`. Then test ingress-app build on branch https://github.com/gravitational/ingress-app/pull/2

```sh
# .../gravitational/ingress-app before changes
$ make tarball
gravity app delete --force --insecure           \
	                            \
	gravitational.io/ingress-app:0.0.1-9-g3849c53
gravitational.io/ingress-app:0.0.1-9-g3849c53 has been deleted
gravity app import --vendor --version=0.0.1-9-g3849c53 \
	--set-image=quay.io/gravitational/debian-tall:stretch      \
	--include=resources --include=registry .
* [1/4] importing directory .
* [2/4] vendoring docker images from .
* [2/4] app import aborted after 3 seconds
[ERROR]: failed to export image layers
	invalid reference format
make: *** [Makefile:22: import] Error 255
```

```sh
# .../gravitational/ingress-app after changes
$ make tarball
gravity app delete --force --insecure           \
	                            \
	gravitational.io/ingress-app:0.0.1-9-g3849c53
gravitational.io/ingress-app:0.0.1-9-g3849c53 has been deleted
gravity app import --vendor --version=0.0.1-9-g3849c53 \
	--set-image=quay.io/gravitational/debian-tall:stretch      \
	--include=resources --include=registry .
* [1/4] importing directory .
* [2/4] vendoring docker images from .
	Still vendoring docker images from . (10 seconds elapsed)
* [3/4] importing application
* [4/4] gravitational.io/ingress-app:0.0.1-9-g3849c53 imported
* [4/4] app import completed in 23 seconds
gravity package export \
	 \
	gravitational.io/ingress-app:0.0.1-9-g3849c53 /home/berd/go/src/github.com/gravitational/ingress-app/build/ingress-app-0.0.1-9-g3849c53.tar.gz
gravitational.io/ingress-app:0.0.1-9-g3849c53 exported to file /home/berd/go/src/github.com/gravitational/ingress-app/build/ingress-app-0.0.1-9-g3849c53.tar.gz
```

bernardjkim committed:
Include image digest and update parsing implementation

bernardjkim committed:
Add GetTag getter function

Returns Digest as a backup

wadells reviewed (changes requested):
What is the behavior when a tag is provided, but it doesn't match the image hash? E.g:

```
k8s.gcr.io/ingress-nginx/controller:v0.49.3@sha256:0444a4a00b3483171266047e89d9f7aace7b22326373ea5df056081d8550fece
```
I know this is a bit out of scope for the changes here, but I'm curious since this touches on it.

It looks like the `String()` behavior vs the `GetTag()` behavior is different.  `String` prefers the digest whereas `GetTag` prefers the Tag.

Noting the above concern, could you add some [reflexivity](https://en.wikipedia.org/wiki/Reflexive_relation) tests? E.g. `input String` -> `DockerImage{}` -> `output String` and then make sure the two strings are equal for all the various inputs.  I'm guessing this will flush out a couple bugs.

bernardjkim committed:
Add example valid registry names

bernardjkim committed:
Tag and Digest cannot be used interchangably

Both Digest and Tag will now be specified when checking if the image
exists in the registry. Digest is not specified when pushing or tagging.

bernardjkim committed:
Add comment for string output format

bernardjkim commented:
No worries. Thanks for the review!
